### PR TITLE
Adding SqsBatchManager interfaces and Refactoring core BatchManager with interfaces

### PR DESCRIPTION
**Pull request by 16lim21**
## Motivation and Context
I added interfaces for the core `BatchManager` to abstract away the private methods in the implementation. I also created the interfaces and class files for the `SqsBatchManager` to prepare for implementation. 

## Description
I added some interfaces in a new `batchmanager` package that is not internal and abstracts away the implementation details for the underlying `DefaultBatchManager`. Customers will therefore interact with the interfaces and classes in the public `batchmanager` package as opposed to the internal one. 

Meanwhile, I modified the `BatchOverrideConfiguration` class to accept optional values, which would be filled in with default values by a core `BatchConfiguration` class. Services that implement a `BatchManager` will not need to rely on this `BatchConfiguration` class, it is just there to ensure default values are provided in case customers implement their own service-specific batch manager. 

Finally, I added some classes and interfaces for the `SqsBatchManager`. The actual implementation has not been finished yet, nor has tests been written. Some code in the `DefaultSqsBatchManager` are just the functions copied over from the Sqs batch manager integration test.

## Testing
Modifications to the core BatchManager are covered by tests in the last PR. No tests have been written yet for the SqsBatchManager but will be done in a separate PR covering implementation details.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License
- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManager.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+package software.amazon.awssdk.core.batchmanager;
+
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.core.internal.batchmanager.DefaultBatchManager;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+
+// TODO: Add javadoc comments. Should I move JavaDoc comments from DefaultBatchManager here? Or keep comments in both files?
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManager.java:23]: We should move the Javadocs from the implementation to the interface

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManager.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+@SdkPublicApi
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManager.java:24]: Protected API? We can make it public in the future if needed

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManager.java (hunk 1 above)
Comment (16lim21) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManager.java]: Moved javadocs from implementation to interface (kept them for implementation variables).

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManager.java (hunk 2 above)
Comment (16lim21) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManager.java]: Changed to a protected API

File: services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsBatchManager.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+         * <p>
+         * This SDK client must be closed by the caller when it is ready to be disposed. The SDK will not close the
+         * client when the BatchManager is closed.
+         *
+         * @param client the client used to send and receive batch messages.
+         * @return a reference to this object so that method calls can be chained together.
+         */
+        Builder client(SqsClient client);
+
+        SqsBatchManager build();
```
[the remark this answers is not in the document]
Comment (16lim21) [reply services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsBatchManager.java:91]: Added javadocs for build() in both SqsBatchManager and SqsAsyncBatchManager.

File: services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsBatchManager.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+@SdkPublicApi
+public interface SqsBatchManager extends SdkAutoCloseable {
+
+    /**
+     * Buffers outgoing SendMessageRequests on the client and sends them as a SendMessageBatchRequest to SQS. Requests are
+     * batched together according to a batchKey and are sent periodically to SQS. If the number of requests for a batchKey
+     * reaches or exceeds the configured max items, then the requests are immediately flushed and the timeout on the periodic
+     * flush is reset.
+     * By default, messages are batched with a maximum batch size of 10. These settings can be customized via the configuration.
```
[the remark this answers is not in the document]
Comment (16lim21) [reply services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsBatchManager.java:40]: Added <p> tag.

File: services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsBatchManager.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.core.batchmanager.BatchOverrideConfiguration;
+import software.amazon.awssdk.services.sqs.SqsClient;
+import software.amazon.awssdk.services.sqs.internal.batchmanager.DefaultSqsBatchManager;
+import software.amazon.awssdk.services.sqs.model.SendMessageRequest;
+import software.amazon.awssdk.services.sqs.model.SendMessageResponse;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+
+@SdkPublicApi
```
[the remark this answers is not in the document]
Comment (16lim21) [reply services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsBatchManager.java:31]: Added javadocs for all public APIs (BatchManager, BatchManagerBuilder, BatchOverrideConfiguration, SqsBatchManager, SqsAsyncBatchManager).

File: services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsBatchManager.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+     * Buffers outgoing SendMessageRequests on the client and sends them as a SendMessageBatchRequest to SQS. Requests are
+     * batched together according to a batchKey and are sent periodically to SQS. If the number of requests for a batchKey
+     * reaches or exceeds the configured max items, then the requests are immediately flushed and the timeout on the periodic
+     * flush is reset.
+     * By default, messages are batched with a maximum batch size of 10. These settings can be customized via the configuration.
+     *
+     * @param message the outgoing SendMessageRequest.
+     * @return a CompletableFuture of the corresponding SendMessageResponse.
+     */
+    CompletableFuture<SendMessageResponse> sendMessage(SendMessageRequest message);
```
[the remark this answers is not in the document]
Comment (16lim21) [reply services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsBatchManager.java:49]: Added "batchable" Apis (only other ones were deleteMessage and changeMessageVisibility).

File: services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsAsyncBatchManager.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+     * batched together according to a batchKey and are sent periodically to SQS. If the number of requests for a batchKey
+     * reaches or exceeds the configured max items, then the requests are immediately flushed and the timeout on the periodic
+     * flush is reset.
+     * <p>
+     * By default, messages are batched with a maximum batch size of 10. These settings can be customized via the configuration.
+     *
+     * @param message the outgoing SendMessageRequest.
+     * @return a CompletableFuture of the corresponding SendMessageResponse.
+     */
+    CompletableFuture<SendMessageResponse> sendMessage(SendMessageRequest message);
```
Comment (zoewangg) [inline services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsAsyncBatchManager.java:48]: Should we throw UnsupportedOperation? That's what we do for the APIs in the low level client interface

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManagerBuilder.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+package software.amazon.awssdk.core.batchmanager;
+
+import java.util.concurrent.ScheduledExecutorService;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.core.internal.batchmanager.BatchAndSend;
+import software.amazon.awssdk.core.internal.batchmanager.BatchKeyMapper;
+import software.amazon.awssdk.core.internal.batchmanager.BatchResponseMapper;
+
+@SdkPublicApi
```
[the remark this answers is not in the document]
Comment (16lim21) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManagerBuilder.java:24]: Makes sense to be a protected API. Should I move the protected API into the internal batchmanager package then?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchOverrideConfiguration.java [hunk 9]
```diff
@@ -96,44 +87,47 @@ public boolean equals(Object o) {
...
             this.maxBatchItems = maxBatchItems;
             return this;
         }
 
+        /**
+         * The maximum amount of time (in milliseconds) that an outgoing call waits for other requests before sending out a batch
+         * request.
+         *
+         * @param maxBatchOpenInMs
+         * @return
```
[the remark this answers is not in the document]
Comment (16lim21) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchOverrideConfiguration.java]: Added docs.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManagerBuilder.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    B scheduledExecutor(ScheduledExecutorService scheduledExecutor);
+
+    /**
+     * Adds a function that defines how requests should be batched together into the appropriate batch response.
+     *
+     * @param batchingFunction the provided function.
+     * @return a reference to this object so that method calls can be chained together.
+     */
+    B batchingFunction(BatchAndSend<RequestT, BatchResponseT> batchingFunction);
```
[the remark this answers is not in the document]
Comment (16lim21) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManagerBuilder.java]: Renamed to batchFunction

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManagerBuilder.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Adds a function that defines how a batch response should be extracted and transformed into its individual responses.
+     *
+     * @param mapResponsesFunction the provided function.
+     * @return a reference to this object so that method calls can be chained together.
+     */
+    B mapResponsesFunction(BatchResponseMapper<BatchResponseT, ResponseT> mapResponsesFunction);
```
[the remark this answers is not in the document]
Comment (16lim21) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManagerBuilder.java]: renamed to responseMapper

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManagerBuilder.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Adds a function that calculates an appropriate batchKey from a given request.
+     *
+     * @param batchKeyMapperFunction the provided function.
+     * @return a reference to this object so that method calls can be chained together.
+     */
+    B batchKeyMapperFunction(BatchKeyMapper<RequestT> batchKeyMapperFunction);
```
[the remark this answers is not in the document]
Comment (16lim21) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManagerBuilder.java]: renamed to batchKeyMapper

File: services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsAsyncBatchManager.java (hunk 7 above)
Comment (16lim21) [reply services/sqs/src/main/java/software/amazon/awssdk/services/sqs/batchmanager/SqsAsyncBatchManager.java]: Default behavior throws unsopportedOperation in SqsBatchManager and SqsAsyncBatchManager

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManagerBuilder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (zoewangg) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/batchmanager/BatchManagerBuilder.java]: No, protected APIs actually should be not be put into the internal package.

**Review by zoewangg — COMMENTED** [review]
Build failed because of a checkstyle error in sqs, otherwise lgtm

**Review by zoewangg — APPROVED** [review]

**Pull request merged**